src/server.ts

```diff
@@ -1,226 +1,385 @@
-	const capabilities = params.capabilities;
-
-	// Does the client support the `workspace/configuration` request?
-	// If not, we fall back using global settings.
-	hasConfigurationCapability = !!(
-		capabilities.workspace && !!capabilities.workspace.configuration
-	);
-	hasWorkspaceFolderCapability = !!(
-		capabilities.workspace && !!capabilities.workspace.workspaceFolders
-	);
-	hasDiagnosticRelatedInformationCapability = !!(
-		capabilities.textDocument &&
-		capabilities.textDocument.publishDiagnostics &&
-		capabilities.textDocument.publishDiagnostics.relatedInformation
-	);
-
-	const result: InitializeResult = {
-		capabilities: {
-			textDocumentSync: TextDocumentSyncKind.Incremental,
-			// Tell the client that this server supports code completion.
-			completionProvider: {
-				resolveProvider: true
-			}
-		}
-	};
-	if (hasWorkspaceFolderCapability) {
-		result.capabilities.workspace = {
-			workspaceFolders: {
-				supported: true
-			}
-		};
-	}
-	return result;
+    const capabilities = params.capabilities;
+    
+    return {
+        capabilities: {
+            textDocumentSync: TextDocumentSyncKind.Incremental,
+            completionProvider: { 
+                resolveProvider: true,
+                triggerCharacters: ['.', ':', '<', '"', '/']
+            },
+            hoverProvider: true,
+            signatureHelpProvider: {
+                triggerCharacters: ['('],
+                retriggerCharacters: [',']
+            },
+            definitionProvider: true,
+            documentFormattingProvider: true,
+            documentSymbolProvider: true,
+            referencesProvider: true,
+            workspaceSymbolProvider: true
+        }
+    } satisfies InitializeResult;
 });
 
-connection.onInitialized(() => {
-	if (hasConfigurationCapability) {
-		// Register for all configuration changes.
-		connection.client.register(DidChangeConfigurationNotification.type, undefined);
-	}
-	if (hasWorkspaceFolderCapability) {
-		connection.workspace.onDidChangeWorkspaceFolders(_event => {
-			connection.console.log('Workspace folder change event received.');
-		});
-	}
+// Configuration Handling
+connection.onDidChangeConfiguration(change => {
+    globalSettings = change.settings.parasailServer || globalSettings;
+    documents.all().forEach(validateDocument);
 });
 
-// The example settings
-interface ExampleSettings {
-	maxNumberOfProblems: number;
+// Document Management
+documents.onDidChangeContent(change => validateDocument(change.document));
+documents.onDidClose(e => documentSettings.delete(e.document.uri));
+
+// Core Validation Logic
+async function validateDocument(document: TextDocument): Promise<void> {
+    const diagnostics: Diagnostic[] = [];
+    const tempFile = createTempFile(document.getText());
+    
+    try {
+        const parser = cp.spawn('interp.csh', [tempFile]);
+        
+        parser.stderr.on('data', (data) => {
+            processDiagnostics(data.toString(), document.uri, diagnostics);
```

Comment:
This might hang up with debugger - haven't checked yet. If that happens, use the `yes` command to autoexit out of the psc debugger.

Reply:
This does seem to happen sometimes. Adding

```
parser.stdin.write("exit");
parser.stdin.end();
```

after the spawn call seems to help on my end